File: 2024/july/javaScript/10.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseParentheses = function(s) {
    const sz = s.length;
    let ans = "";

    const reverseAndReplace = function( st , end , ans){
        ++end;
        const before = ans.substring(0 , st);
        let toChange = ans.substring(st , end);
        const after = ans.substring(end);
        toChange = toChange.split('').reverse().join('');
        return before + toChange+ after;
    };
    let pre = [];
    let cnt = 0;
    let st = [];
    for (let i = 0; i < sz; ++i){
        const c = s[i];
        if (c == '('){
            st.push(i);
            pre[i] = ++cnt;
        }else if (c == ')'){
            pre[i] = ++cnt;
            let popedElement = st.pop();
            ans = reverseAndReplace(popedElement-pre[popedElement]+1 , i-pre[i] , ans);
             
        }else{
           ans+= c;
        }
        pre[i] = cnt;
    }
    return ans;
    
};

File: 2024/july/java/10.java
class Solution {
    private void reverseAndReplace(int st , int end , StringBuilder ans){
        ++end;
        ans.replace(st, end, new StringBuilder(ans.substring(st, end)).reverse().toString());
    }
    public String reverseParentheses(String s) {
        var st = new Stack<Integer>();
        var ans = new StringBuilder();
        int cnt = 0;
        final int sz = s.length();
        var pre = new int [sz];
      
        for (int i = 0; i < sz; ++i){
            char c = s.charAt(i);
            if (c == '('){
                st.push(i);
                pre[i] = ++cnt;
            }else if (c == ')'){
                pre[i] = ++cnt;
                int popedElement = st.pop();
               reverseAndReplace(popedElement-pre[popedElement]+1 , i-pre[i] , ans);
             
            }else{
                ans.append(c);
            }
            pre[i] = cnt;
        }
        return ans.toString();
        
    }
}
